# SQL

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Ваша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. Ваша первая задача как аналитика — проанализировать базу данных.
В ней — информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

### Описание данных

**Таблица `books`**

Содержит данные о книгах:

- `book_id` — идентификатор книги;
- `author_id` — идентификатор автора;
- `title` — название книги;
- `num_pages` — количество страниц;
- `publication_date` — дата публикации книги;
- `publisher_id` — идентификатор издателя.

**Таблица `authors`**

Содержит данные об авторах:

- `author_id` — идентификатор автора;
- `author` — имя автора.

**Таблица `publishers`**

Содержит данные об издательствах:

- `publisher_id` — идентификатор издательства;
- `publisher` — название издательства;

**Таблица `ratings`**

Содержит данные о пользовательских оценках книг:

- `rating_id` — идентификатор оценки;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, оставившего оценку;
- `rating` — оценка книги.

**Таблица `reviews`**

Содержит данные о пользовательских обзорах на книги:

- `review_id` — идентификатор обзора;
- `book_id` — идентификатор книги;
- `username` — имя пользователя, написавшего обзор;
- `text` — текст обзора.

### Задания

- Посчитайте, сколько книг вышло после 1 января 2000 года;
- Для каждой книги посчитайте количество обзоров и среднюю оценку;
- Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры;
- Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками;
- Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import text, create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{user}:{pwd}@{host}:{port}/{db}'.format(**db_config)

# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'})

# чтобы выполнить SQL-запрос, используем Pandas
query = '''SELECT * FROM books LIMIT 5'''

con=engine.connect()

pd.io.sql.read_sql(sql=text(query), con = con)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Изучим данные в каждой таблице

Таблица `books`

In [3]:
query = '''SELECT * FROM books LIMIT 5'''

books = pd.io.sql.read_sql(query, con = engine)

display(books.head())
print()
display(books.info())

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5 entries, 0 to 4
Data columns (total 6 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   book_id           5 non-null      int64 
 1   author_id         5 non-null      int64 
 2   title             5 non-null      object
 3   num_pages         5 non-null      int64 
 4   publication_date  5 non-null      object
 5   publisher_id      5 non-null      int64 
dtypes: int64(4), object(2)
memory usage: 368.0+ bytes


None

Таблица `ratings`

In [4]:
query = '''SELECT * FROM ratings'''

ratings = pd.io.sql.read_sql(query, con = engine)

display(ratings.head())
print()
display(ratings.info())

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6456 entries, 0 to 6455
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   rating_id  6456 non-null   int64 
 1   book_id    6456 non-null   int64 
 2   username   6456 non-null   object
 3   rating     6456 non-null   int64 
dtypes: int64(3), object(1)
memory usage: 201.9+ KB


None

Таблица `authors`

In [5]:
query = '''SELECT * FROM authors'''

authors = pd.io.sql.read_sql(query, con = engine)

display(authors.head())
print()
display(authors.info())

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 636 entries, 0 to 635
Data columns (total 2 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   author_id  636 non-null    int64 
 1   author     636 non-null    object
dtypes: int64(1), object(1)
memory usage: 10.1+ KB


None

Таблица `reviews`

In [6]:
query = '''SELECT * FROM reviews'''

reviews = pd.io.sql.read_sql(query, con = engine)

display(reviews.head())
print()
display(reviews.info())

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 2793 entries, 0 to 2792
Data columns (total 4 columns):
 #   Column     Non-Null Count  Dtype 
---  ------     --------------  ----- 
 0   review_id  2793 non-null   int64 
 1   book_id    2793 non-null   int64 
 2   username   2793 non-null   object
 3   text       2793 non-null   object
dtypes: int64(2), object(2)
memory usage: 87.4+ KB


None

### Задание 1: Посчитайте, сколько книг вышло после 1 января 2000 года.

In [7]:
query = '''SELECT COUNT(DISTINCT book_id)
FROM books 
WHERE publication_date > '2000-01-01'
    '''
pd.io.sql.read_sql(query, con = engine)

count
0    819

### Задание 2: Для каждой книги посчитайте количество обзоров и среднюю оценку.

In [8]:
query = '''SELECT books.book_id, title, COUNT(DISTINCT reviews.book_id), ROUND(AVG(rating),2) AS avg_rating
FROM books 
LEFT JOIN reviews ON books.book_id = reviews.book_id
LEFT JOIN ratings ON books.book_id = ratings.book_id
GROUP BY books.book_id
ORDER BY avg_rating DESC , count
'''

pd.io.sql.read_sql(query, con = engine)


book_id                                              title  count  \
0        672   The Cat in the Hat and Other Dr. Seuss Favorites      0   
1        732  The Ghost Map: The Story of London's Most Terr...      1   
2        182                                         Dead Souls      1   
3        169  Crucial Conversations: Tools for Talking When ...      1   
4        321                                Homage to Catalonia      1   
..       ...                                                ...    ...   
995      915  The World Is Flat: A Brief History of the Twen...      1   
996      202                                      Drowning Ruth      1   
997      371                                              Junky      1   
998      316                  His Excellency: George Washington      1   
999      303                               Harvesting the Heart      1   

     avg_rating  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 4 columns]

### Задача 3: Определите издательство, которое выпустило наибольшее число книг толще 50 страниц — так вы исключите из анализа брошюры.

In [9]:
query = '''SELECT  COUNT(b.book_id) AS count_books, p.publisher
FROM books b
LEFT JOIN publishers p
ON b.publisher_id=p.publisher_id
WHERE num_pages > 50
GROUP BY publisher
ORDER BY count_books DESC
LIMIT 10'''

pd.io.sql.read_sql(query, con = engine)

count_books                  publisher
0           42              Penguin Books
1           31                    Vintage
2           25   Grand Central Publishing
3           24           Penguin Classics
4           19           Ballantine Books
5           19                     Bantam
6           17                    Berkley
7           14         St. Martin's Press
8           14              Berkley Books
9           13  William Morrow Paperbacks

### Задача 4: Определите автора с самой высокой средней оценкой книг — учитывайте только книги с 50 и более оценками.

In [10]:
query = '''
WITH rt AS 
(SELECT book_id
FROM ratings
GROUP BY book_id
HAVING COUNT(rating_id) >= 50)

SELECT a.author,  ROUND(AVG(r.rating),2) AS avg_rating
FROM books AS b

 JOIN rt ON b.book_id = rt.book_id

 JOIN authors AS a ON b.author_id = a.author_id

 JOIN ratings AS r ON rt.book_id = r.book_id

GROUP BY a.author
ORDER BY avg_rating DESC'''

pd.io.sql.read_sql(query, con = engine)

author  avg_rating
0                          J.K. Rowling/Mary GrandPré        4.29
1                   Markus Zusak/Cao Xuân Việt Khương        4.26
2                                      J.R.R. Tolkien        4.25
3                                   Louisa May Alcott        4.19
4                                        Rick Riordan        4.08
5                                     William Golding        3.90
6                                       J.D. Salinger        3.83
7   William Shakespeare/Paul Werstine/Barbara A. M...        3.79
8            Paulo Coelho/Alan R. Clarke/Özdemir İnce        3.79
9                                          Lois Lowry        3.75
10                                          Dan Brown        3.74
11            George Orwell/Boris Grabnar/Peter Škerl        3.73
12                                    Stephenie Meyer        3.66
13                                     John Steinbeck        3.62

### Задача 5: Посчитайте среднее количество обзоров от пользователей, которые поставили больше 48 оценок.

In [11]:
query = '''
WITH user_rating AS 
(SELECT rt.username AS rating_username, COUNT(rating_id) AS ratings_count
FROM ratings rt
GROUP BY rt.username
HAVING COUNT(rating_id) > 48),
        
 user_review AS 
(SELECT rv.username AS review_username, COUNT(review_id) AS review_count
FROM reviews rv
GROUP BY rv.username)

SELECT AVG(review_count) AS avg_review_count
FROM user_rating
JOIN user_review ON user_rating.rating_username = user_review.review_username
     
ORDER BY avg_review_count  DESC'''

pd.io.sql.read_sql(query, con = engine)

avg_review_count
0              24.0

Вывод:
- 819 книг вышло после 01 января 2020г;
- У книги "The Cat in the Hat and Other Dr. Seuss Favorites" нет ни одного обзора, но высокий рейтинг -5;
- Издательство "Penguin Books" выпустило наибольшее число книг толще 50 страниц - 42 книги;
- Автор с самой высокой средней оценкой книг по заданому критирею явлется J.K. Rowling/Mary GrandPré, имеет средний рейтинг 4.29;
- Cреднее количество обзоров от пользователей, которые поставили больше 48 оценок - 24.